src/Authentication/Authentication/custom/common/Permissions.ps1

```diff
@@ -0,0 +1,102 @@
+# ------------------------------------------------------------------------------
+#  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+# ------------------------------------------------------------------------------
+
+Set-StrictMode -Version 6.0
+
+$Permissions_msGraphApplicationId = '00000003-0000-0000-c000-000000000000'
+
+# Entry point to the state of the Permissions 'class'
+$_permissions = $null
+
+# These '_'-prefixed functions are provided for tests to simulate the initial state of the class
+# as well as providing visibility into its state for deeper validations
+
+function _Permissions_Initialize {
+    # Testing note: Pester alters runtime behavior so that variables defined at script
+    # scope do not actually show up at script scope via 'script:' at runtime (!), so
+    # we'll just use the Get-Variable command to explicitly obtain the variable object for it
+    $permissionsVariable = Get-Variable _permissions
+
+    # This structure models the state of the permissions class
+    $permissionsVariable.Value = [PSCustomObject] @{
+        msGraphPermissionsRequestUri = "https://graph.microsoft.com/v1.0/servicePrincipals?`$filter=appId eq '$Permissions_msGraphApplicationId'"
+        msGraphServicePrincipal = $null
+        isFromInvokeMgGraphRequest = $false
+    }
+}
+
+function _Permissions_State {
+    $_permissions
+}
+
+function Permissions_GetPermissionsData([bool] $online) {
+    # Make a REST request to MS Graph to get the permissions data from the Microsoft Graph service principal
+    if ( $online -or ! $_permissions.msGraphServicePrincipal -or ! $_permissions.isFromInvokeMgGraphRequest ) {
+        try {
+            $restResult = Invoke-MgGraphRequest -method GET $_permissions.msGraphPermissionsRequestUri
+
+            if ( $restResult ) {
+                $_permissions.msGraphServicePrincipal = $restResult | Select-Object -ExpandProperty value
+                $_permissions.isFromInvokeMgGraphRequest = $true
+            }
+        } catch [System.Management.Automation.ValidationMetadataException], [System.Net.Http.HttpRequestException] {
+            # We can't get the data from MS Graph, so just use a local static (possibly stale) copy
+            $_permissions.msGraphServicePrincipal = Get-Content $PSScriptRoot/MSGraphServicePrincipalPermissions.json | Out-String | ConvertFrom-Json
+            if ( $online ) {
+                throw
+            }
```

Comment:
To be memory efficient, you should throw the exception first (when `-Online` is true) before reading the file.
```suggestion
            if ( $online ) {
                throw
            }
            # We can't get the data from MS Graph, so just use a local static (possibly stale) copy
            $_permissions.msGraphServicePrincipal = Get-Content $PSScriptRoot/MSGraphServicePrincipalPermissions.json | Out-String | ConvertFrom-Json
```